### PR TITLE
Fix iOS build breakages by adding http::Request to iOS build

### PR DESCRIPTION
**Pull request by plubius**
Still needed to add a bunch of #if !MOBILAPP to disable all of the Online server's unit testing classes so that the iOS app won't crash when opening a document.

TODO: the iOS app will get stuck in an infinite loop when opening a document on iOS in http::Request::readData().

### CODE REVIEW

File: wsd/COOLWSD.cpp [hunk 1]
```diff
@@ -3575,7 +3577,6 @@ class PrisonerRequestDispatcher final : public WebSocketHandler
 #else
             pid_t pid = 100;
             std::string jailId = "jail";
-            LOG_ASSERT_MSG(socket->getInBuffer().empty(), "Unexpected data in prisoner socket");
```
Comment (plubius) [inline wsd/COOLWSD.cpp:3578]: Note: on mobile, the socket is always empty as JavaScript messages come over a separate "fake socket" class.

File: wsd/DocumentBroker.hpp [hunk 2]
```diff
@@ -1460,7 +1460,7 @@ class DocumentBroker : public std::enable_shared_from_this<DocumentBroker>
     const bool _alwaysSaveOnExit;
 
     // Last member.
-#ifdef ENABLE_DEBUG
```
Comment (plubius) [inline wsd/DocumentBroker.hpp:1463]: Note: #ifdef ENABLE_DEBUG is always true as config.in will always define it. In a production build config.in contains the following:

#define ENABLE_DEBUG 0

Since _unitWsd was always instantiated in Online builds, production or not, I removed the #ifdef ENABLE_DEBUG.

File: net/WebSocketHandler.hpp [hunk 3]
```diff
@@ -45,7 +45,7 @@ class WebSocketHandler : public ProtocolHandlerInterface
     const bool _isClient;
 
     // Last member.
-#ifdef ENABLE_DEBUG
```
Comment (plubius) [inline net/WebSocketHandler.hpp:48]: Note: #ifdef ENABLE_DEBUG is always true as config.in will always define it. In a production build config.in contains the following:

#define ENABLE_DEBUG 0

Since _unit was always instantiated in Online builds, production or not, I removed the #ifdef ENABLE_DEBUG.

Comment (mmeeks) [reply net/WebSocketHandler.hpp:48]: great catch =)

File: common/Session.cpp [hunk 4]
```diff
@@ -254,13 +254,15 @@ void Session::handleMessage(const std::vector<char> &data)
 {
     try
     {
+#if !MOBILEAPP
```
Comment (mmeeks) [inline common/Session.cpp:257]: Soo ... I suggest that we guard these UnitBase.get() calls with:

UnitBase::isUnitTesting() && ...

And then I think we don't get any crashes; and we don't have any extra compile conditionals =)

File: net/WebSocketHandler.hpp [hunk 5]
```diff
@@ -83,7 +83,7 @@ class WebSocketHandler : public ProtocolHandlerInterface
 #endif
         _shuttingDown(false)
         , _isClient(isClient)
-#ifdef ENABLE_DEBUG
+#if !MOBILEAPP
         , _unit(UnitBase::get())
```
Comment (mmeeks) [inline net/WebSocketHandler.hpp:87]: This boggles the mind - why do we have a member in each instance that stores the result of a static function - that seems unbelievable =)

Ash - what is going on here:

commit a89ed3b5350a55d265d5ea0131f3a342e28384ff
Author: Ashod Nakashian <ashod.nakashian@collabora.co.uk>

And why can't we use the global variable; that's really under-explained; also see Patrick's point on the #ifdef FOO.

Ideally we would have a git hook to check for #ifdefs and stop them creeping in in error =)

**Review by mmeeks — COMMENTED** [review]
I know it's not your fault that there are lots of conditionals [!]; but I'm really trying to not make the maintainability problem worse here =)

I think we can instead use the method UnitBase::isUnitTesting.

If you must get it in as-is fine; but we badly need to be converging not diverging code here for mobile I think.

Possibly (or probably) one of the problems here is the UnitWSD vs. UnitKit when they're linked into the same process, which makes life less elegant, but the UnitBase::isUnitTesting() should be fine either way to guard those accessess.

Thanks!

File: common/Session.cpp (hunk 4 above)
Comment (plubius) [reply common/Session.cpp:257]: OK. That founds like a reasonable approach. I'll see if I can implement that over the next day or two.

Is it safe to assume that any UnitWSD calls can also be guarded with UnitBase::isUnitTesting()?

**Comment by plubius** [comment]
An update on the "the iOS app will not accept text input at all" bug that I am seeing: it seems to only be limited to Writer and Impress documents but not Calc. Also, arrow keys and formatting toolbar buttons both work, I just can't enter any new text or delete text via key events.

File: net/HttpRequest.cpp [hunk 6]
```diff
@@ -307,7 +307,13 @@ int64_t Request::readData(const char* p, const int64_t len)
     if (_stage == Stage::Header)
     {
         // First line is the status line.
+#if !MOBILEAPP
         if (p == nullptr || len < MinRequestHeaderLen)
+#else
+        // Fix infinite loop on mobile by skipping the minimum request header
```
Comment (Ashod) [inline net/HttpRequest.cpp:313]: Wouldn't it make more sense to define `MinRequestHeaderLen` differently for mobile, instead of explicitly short-circuiting here?

But even that seems odd. I mean, this is generic HTTP parser. Why would we blanket disable it on Mobile? Doesn't make sense. Surely we use HTTP on Mobile! Perhaps we don't use sockets for certain cases, but disabling such independent code would make it harder to use this code anywhere in Mobile at all.

File: net/WebSocketHandler.hpp [hunk 7]
```diff
@@ -669,10 +668,10 @@ class WebSocketHandler : public ProtocolHandlerInterface
     /// 0 for closed socket, and -1 for other errors.
     int sendMessage(const char* data, const size_t len, const WSOpCode code, const bool flush) const
     {
-        if (!Util::isFuzzing())
+        if (UnitBase::isUnitTesting() && !Util::isFuzzing())
```
Comment (Ashod) [inline net/WebSocketHandler.hpp:671]: We shouldn't need to check `isUnitTesting()`, because the default UnitBase object has no-op implementation. But perhaps we don't want to instantiate them on Mobile? Is that the idea here?

File: wsd/COOLWSD.cpp (hunk 1 above)
Comment (Ashod) [reply wsd/COOLWSD.cpp:3578]: Great! So why is asserting that it *is* empty wrong?

File: wsd/DocumentBroker.cpp [hunk 8]
```diff
@@ -149,13 +149,14 @@ DocumentBroker::DocumentBroker(ChildType type, const std::string& uri, const Poc
...
     , _alwaysSaveOnExit(COOLWSD::getConfigValue<bool>("per_document.always_save_on_exit", false))
-#ifdef ENABLE_DEBUG
-    , _unitWsd(UnitWSD::get())
-#endif
+    , _unitWsd(nullptr)
 {
     assert(!_docKey.empty());
     assert(!COOLWSD::ChildRoot.empty());
 
+    if (UnitWSD::isUnitTesting())
```
Comment (Ashod) [inline wsd/DocumentBroker.cpp:157]: Can we please do this in the initialization list, and have a const pointer?
```
, _unitWsd(UnitWSD::isUnitTesting() ? nullptr : &UnitWSD::get())
```

(Same elsewhere.)

File: wsd/DocumentBroker.hpp [hunk 9]
```diff
@@ -1460,12 +1460,10 @@ class DocumentBroker : public std::enable_shared_from_this<DocumentBroker>
...
     /// The UnitWSD instance. We capture it here since
     /// this is our instance, but the test framework
     /// has a single global instance via UnitWSD::get().
-    UnitWSD& _unitWsd;
-#endif
+    UnitWSD* _unitWsd;
```
Comment (Ashod) [inline wsd/DocumentBroker.hpp:1466]: `UnitWSD* const _unitWsd;`

(Same elsewhere.)

File: net/WebSocketHandler.hpp (hunk 5 above)
Comment (Ashod) [reply net/WebSocketHandler.hpp]: Good question! Not obvious, but we now support running multiple tests in the same process. To do that, we have to create a new instance of the UnitX. The problem is that we have all sorts of threads and async logic and re-using the static version sends events/data to the wrong test (when we "switch" from one test to the next).

Multi-test support makes it far easier to add new variations on a given test-case (in the same unit-test files, no Makefile changes), faster build (no new translation units), and faster runtime (same process, same Kit fork). It's wins all around!

(The fact that `ENABLE_DEBUG` was always true is neither here nor there; tests are _explicitly_ disabled outside of debug builds.)

File: wsd/DocumentBroker.hpp (hunk 2 above)
Comment (Ashod) [reply wsd/DocumentBroker.hpp:1463]: Shouldn't we fix `ENABLE_DEBUG` instead? (Test cannot run in non-debug builds.)

I'd rather we keep this and fix configure.ac. Technically, it's correct that we need test code in debug builds only.

**Review by Ashod — COMMENTED** [review]
Thanks @plubius. The approach in this PR looks more promising than the previous attempts. Thanks for persevering!

In addition to @mmeeks's notes, I've left some further suggestions. I hope they make sense, but feel free to ask if you think otherwise.

Thanks again!

File: net/WebSocketHandler.hpp (hunk 7 above)
Comment (mmeeks) [reply net/WebSocketHandler.hpp:671]: @Ashod ah - quite possibly we are just failing to do the no-op instantiation on mobile - *but* bear in mind that we have both Kit and WSD in the same process - so probably this one global that is both of these is a problem (or not). If you could dig out the code to do the no-op initialization for Patrick that would be much appreciated - Thanks ! =)

File: net/HttpRequest.cpp (hunk 6 above)
Comment (mmeeks) [reply net/HttpRequest.cpp:313]: We shouldn't use HTTP on Mobile - we should just push the messages through without wrapping in a websocket container for performance & slickness.

**Comment by mmeeks** [comment]
> An update on the "the iOS app will not accept text input at all" bug that I am seeing: it seems to only be limited to Writer and Impress documents but not Calc. Also, arrow keys and formatting toolbar buttons both work, I just can't enter any new text or delete text via key events.

@mcecchetti - any chance you can have a quick look at this. Patrick - Marco has been re-working the input methods - I guess you building for iOS is not possible - but if you could read around the code and offer some suggestions that would be great =) thanks !

File: net/WebSocketHandler.hpp (hunk 7 above)
Comment (Ashod) [reply net/WebSocketHandler.hpp:671]: `UnitBase::init()` has a fallback case (when no unit-lib path is given) that creates a `UnitWSD` or `UnitKit` instance.

E.g.:
```
            rememberInstance(UnitType::Wsd, new UnitWSD("UnitWSD"));
```

**Comment by plubius** [comment]
> @mcecchetti - any chance you can have a quick look at this. Patrick - Marco has been re-working the input methods - I guess you building for iOS is not possible - but if you could read around the code and offer some suggestions that would be great =) thanks !

So my first guess is that this might be a Safari/WKWebKit thing. I can confirm that the JS code is not sending some key events as messages to the fake socket JS object. For example, in Calc if I type in the cell's input field in the toolbar, I see events like the following sent to the fake socket which is expected:

2023-06-15 09:10:22.540651 -0400 [ main ] TRC  To Online: textinput id=0 text=j| DocumentViewController.mm:288

But in Writer, I see such events for arrow keys, but no events are written to the fake socket for regular ASCII character text input.

File: wsd/COOLWSD.cpp (hunk 1 above)
Comment (plubius) [reply wsd/COOLWSD.cpp:3578]: > Great! So why is asserting that it _is_ empty wrong?

Because that socket is *always* empty. This code is called only once when a document is first opened and I doubt that an actual C socket is ever created so the socket buffer will always be empty.

I would guess that the only reason that iOS invokes this code is that is a side-effect of bootstrapping LibreOffice and starting the LOKit queue. After this code is called, iOS immediately tells the JS code to set app.Socket to a fake socket. The Online server components are never sent messages directly on mobile.

Instead, on iOS the fake socket routes messages to -[DocumentViewController userContentController:didReceiveScriptMessage:] in ios/Mobile/DocumentViewController.mm. This code forwards the message directly to LOKit, bypassing the Online server completely.

File: wsd/DocumentBroker.cpp (hunk 8 above)
Comment (plubius) [reply wsd/DocumentBroker.cpp]: Done

File: wsd/DocumentBroker.hpp (hunk 9 above)
Comment (plubius) [reply wsd/DocumentBroker.hpp]: Done

File: common/Session.cpp (hunk 4 above)
Comment (plubius) [reply common/Session.cpp]: Done

File: net/HttpRequest.cpp [hunk 10]
```diff
@@ -116,7 +116,7 @@ static inline int64_t findEndOfToken(const char* p, int64_t off, int64_t len)
 
 int64_t Header::parse(const char* p, int64_t len)
 {
-    LOG_TRC("Parsing header given " << len << " bytes: " << std::string(p, std::min(len, 80L)));
+    LOG_TRC("Parsing header given " << len << " bytes: " << std::string(p, std::min(len, (int64_t)80L)));
```
[the remark this answers is not in the document]
Comment (plubius) [reply net/HttpRequest.cpp]: Done

File: net/WebSocketHandler.hpp (hunk 7 above)
Comment (plubius) [reply net/WebSocketHandler.hpp:668]: I would vote against this for a couple of reasons:

1. We would probably need to pull more Online server code into the Xcode build which will likely bring in more code that doesn't build or run on iOS. AFAICT, there no tests for iOS at this time so sounds like a lot of work for no immediate gain.
2. iOS is a very resource constrained environment. Adding a bunch of code that just do nothing on iOS sounds like a very bad idea given how much effort was spent in March to make the iOS app not crash due to out of memory errors and performance bottlenecks.

By all means, if someone implements unit tests that can on iOS, go ahead and add them but can you do that in a separate pull request? Feel free to add me as a reviewer and I can see if it has any problems building or running.

**Review by mmeeks — APPROVED** [review]
Looks beautiful to me - thanks Patrick ! =)

**Comment by mmeeks** [comment]
CI runs builds and tests on Linux - that's fine =)

**Comment by plubius** [comment]
Thanks so much!

**Comment by mcecchetti** [comment]
> So my first guess is that this might be a Safari/WKWebKit thing. I can confirm that the JS code is not sending some key events as messages to the fake socket JS object. For example, in Calc if I type in the cell's input field in the toolbar, I see events like the following sent to the fake socket which is expected:
> 
> 2023-06-15 09:10:22.540651 -0400 [ main ] TRC To Online: textinput id=0 text=j| DocumentViewController.mm:288
> 
> But in Writer, I see such events for arrow keys, but no events are written to the fake socket for regular ASCII character text input.

I performed some test: when Online is loaded in Safari iOS I'm able to see `textinput` events sent from the client. Typing is working fine. I checked the `TextInput` class: there are only a few code fragments specific for the iOS app, and they didn't seem to have been modified recently.  Could it be a focus issue ? I mean maybe there is something preventing the editable area to get focused. Are you using a virtual or hw-keyboard ?
A last question: does text input in Calc work both by using the formula bar and by typing directly into the cell ?

**Comment by plubius** [comment]
> I performed some test: when Online is loaded in Safari iOS I'm able to see `textinput` events sent from the client. Typing is working fine. I checked the `TextInput` class: there are only a few code fragments specific for the iOS app, and they didn't seem to have been modified recently. Could it be a focus issue ? I mean maybe there is something preventing the editable area to get focused. Are you using a virtual or hw-keyboard ? A last question: does text input in Calc work both by using the formula bar and by typing directly into the cell ?

Thank you for taking a look at this. I am seeing this with both hardware and virtual keyboards. Typing directly into a cell does not work, but in both Writer and Calc, I can input text into JS dialogs. So, it looks like only the document area is affected.

I'm not sure if focus is part of the problem. I am seeing text messages come through from the JS code for "action "keys (e.g. arrows, tab, key shortcuts like Command-C/V/Z, etc.), but not for actual text.

Since this works on Safari, when I get some time I will probably need to step through the code and see if the JS code is hitting any special "is mobile" or "is iOS" code.

**Comment by mcecchetti** [comment]
> Since this works on Safari, when I get some time I will probably need to step through the code and see if the JS code is hitting any special "is mobile" or "is iOS" code.

I would suggest to give a glance at this piece of code: https://github.com/CollaboraOnline/online/blob/11767353b3e15063da17eba66b0ec3b14a5e08fc/browser/src/layer/marker/TextInput.js#LL218C32-L218C32

Please note the `ThisIsTheiOSApp` flag. Though, there is obvious reason, it shouldn't work correctly for the iOS app.

**Comment by plubius** [comment]
> Please note the `ThisIsTheiOSApp` flag. Though, there is no obvious reason, it shouldn't work correctly for the iOS app.

Thank you for the code pointer. I put some console.warn() calls in that function and the good news is that every pass through that function appears to be invoking this._setAcceptInput(true).

I am seeing the JS code log all incoming key events so I will start there and walk through the code until I find the branch point.

**Comment by plubius** [comment]
OK. I think I know what is happening: the main text area is not receiving "input" events in the _onInput handler in browser/src/map/handler/Map.Keyboard.js.

I suspect that it is a focus issue. If the text area doesn't have focus
as when I use an external keyboard, it works fine entering text in the "Search" field in Writer's status bar but not in the main document area.

The good news is that without an external keyboard, tapping in the main document area displays the virtual keyboard and text input works perfectly. Even issue #5837 is fixed.

I'll debug this more when I get some time.

**Comment by mcecchetti** [comment]
> The good news is that without an external keyboard, tapping in the main document area displays the virtual keyboard and text input works perfectly. Even issue #5837 is fixed.

Well, that's a great news that with the virtual keyboard typing is working :-)
I would point out the following code fragments:
https://github.com/CollaboraOnline/online/blob/1cf561d0ef2b9b8130bf077355700c93e0898f7d/browser/src/layer/marker/TextInput.js#L118-L130

and:
https://github.com/CollaboraOnline/online/blob/1cf561d0ef2b9b8130bf077355700c93e0898f7d/browser/html/cool.html.m4#L109-L124

Since you have problem with hw keyboard only, it could be worth checking that the above workaround is still working.
Maybe giving a glance to this commit https://github.com/CollaboraOnline/online/commit/3df718aac72138849ecafadc13f0207c3ec52283 could be helpful too.

**Comment by plubius** [comment]
Thank you for the code pointers. I can confirm that the JS window event listeners are being added via the native keyboard detection code. But after much debugging, I found that by using the debug styles for the "clipboard-area" div in _setupStyles in TextInput.js, I could see that the div acts as if it is not editable.

With the debug styles, you can see the div's cursor blink and text is added to the div when running on an iPad with only a virtual keyboard. However, when I run on macOS (from Xcode) with only the macOS keyboard, there is no cursor and no text ever added.

I suspect (hope) that this is just bugs in Apple's "run iOS app on macOS Silicon" feature. So I will pickup a bluetooth keyboard for my iPad later today and see if I see the same text input problem on an actual iPad.

Nevertheless, I finally was able to get the "clipboard-area" div  to accept input when running the iOS app on macOS with the patch below. Strangely, the "-webkit-user-select: none" style for the parent pane of the div blocks input to the div even though the div has focus and is the activeElement: [iOS_running_on_macOS.txt](https://github.com/CollaboraOnline/online/files/11801163/iOS_running_on_macOS.txt)

TL;DR The switch from a textarea to a div for text input acts oddly when running an iOS app on macOS. Still need to determine if this odd behavior occurs on a real iOS device with a keyboard.

**Comment by plubius** [comment]
Good news! The weird key entry bug I have been posting about does *not* occur on my iPad with a bluetooth keyboard so this is just a but in Apple's "run iOS app on macOS" feature. The iOS app is not shipped as a "Mac Catalyst" compatible app. I only use it to speed up my build and test cycle so I don't think we need to worry about this bug.

I can also confirm that issue #5837 is fixed for me in 23.05 with both a virtual keyboard and a bluetooth keyboard. Also, the native input handler popups (e.g. the accented character list when you press and hold a vowel) works properly and the popup is placed in the proper position near the cursor when using an external keyboard.

Overall, the switch to a <div> for key input in 23.05 appears to work much smoother than the <textarea> used in 22.05.

**Comment by mcecchetti** [comment]
> Good news! The weird key entry bug I have been posting about does _not_ occur on my iPad with a bluetooth keyboard so this is just a but in Apple's "run iOS app on macOS" feature. The iOS app is not shipped as a "Mac Catalyst" compatible app. I only use it to speed up my build and test cycle so I don't think we need to worry about this bug.

Really a good news!
And a really weird bug. It reminded me when I hit a mouse event issue that occurred only when Chrome Development Tools panel was open. At least there is nothing to fix. :-)

Nice to hear that the switch to content editable `div` has solved the emoji issue.
